HomeWorks/HW5/ExpressionTree/ExpressionParser.cs

```diff
@@ -0,0 +1,140 @@
+        }
+
+        char ch = (char)next;
+        if (ch == '(')
+        {
+            // Потребляем открывающую скобку
+            reader.Read();
+            SkipWhiteSpace();
+
+            int opInt = reader.Read();
+            if (opInt == -1)
+            {
+                throw new ParseException("Неожиданный конец входных данных при ожидании оператора.");
+            }
+            char op = (char)opInt;
+            if (op != '+' && op != '-' && op != '*' && op != '/')
+            {
+                throw new ParseException($"Некорректный оператор '{op}'.");
+            }
+
+            Expression left = ParseExpression();
+            Expression right = ParseExpression();
+
+            SkipWhiteSpace();
+            int closing = reader.Read();
+            if (closing != ')')
+            {
+                throw new ParseException("Ожидалась закрывающая скобка ')'.");
+            }
+            return new BinaryOperationExpression(op, left, right);
+        }
+        else
+        {
+            return ParseNumber();
+        }
+    }
+
+    /// <summary>
+    /// Разбирает число (целое) из входных данных.
+    /// </summary>
+    /// <returns>Объект <see cref="NumberExpression"/> с разобранным числом.</returns>
+    /// <exception cref="ParseException">Выбрасывается при ошибках разбора числа.</exception>
+    private NumberExpression ParseNumber()
+    {
+        SkipWhiteSpace();
+        StringBuilder sb = new StringBuilder();
+        int next = reader.Peek();
+        if (next == -1)
+        {
+            throw new ParseException("Неожиданный конец входных данных при ожидании числа.");
+        }
+
+        char ch = (char)next;
+        if (ch == '-' || char.IsDigit(ch))
+        {
+            sb.Append((char)reader.Read());
+            while (true)
+            {
+                next = reader.Peek();
+                if (next == -1)
+                {
+                    break;
+                }
+                ch = (char)next;
+                if (!char.IsDigit(ch))
+                {
+                    break;
+                }
+                sb.Append((char)reader.Read());
+            }
+        }
+        else
+        {
+            throw new ParseException($"Неожиданный символ '{ch}' при ожидании числа.");
+        }
+
+        if (int.TryParse(sb.ToString(), out int value))
+        {
+            return new NumberExpression(value);
+        }
+        else
+        {
+            throw new ParseException($"Некорректный формат числа: {sb}");
+        }
+    }
+
+    /// <summary>
+    /// Пропускает пробельные символы во входных данных.
+    /// </summary>
+    private void SkipWhiteSpace()
+    {
+        while (true)
+        {
+            int next = reader.Peek();
+            if (next == -1)
+            {
+                break;
+            }
+            if (!char.IsWhiteSpace((char)next))
+            {
+                break;
+            }
```

Comment:
```suggestion
            if (next == -1 || !char.IsWhiteSpace((char)next))
            {
                break;
            }
```
Выражение ведь слева направо вычисляется